scalafix/rules/src/main/scala/migrate/MigrationRule.scala

```diff
@@ -0,0 +1,153 @@
+import scala.util.{Failure, Success}
+
+class MigrationRule(global: ScalafixGlobal) extends SemanticRule("MigrationRule") {
+  override def description: String = "infer types and show synthetics"
+
+  def this() = this(null)
+
+  override def withConfiguration(config: Configuration): Configured[Rule] = {
+    if (config.scalacClasspath.isEmpty) {
+      Configured.error(s"config.scalacClasspath should not be empty")
+    } else {
+      val global = CompilerService.newGlobal(config.scalacClasspath, config.scalacOptions)
+      global match {
+        case Success(settings) => Configured.ok(new MigrationRule(new ScalafixGlobal(settings, new StoreReporter, Map())))
+        case Failure(exception) => Configured.error(exception.getMessage)
+      }
+    }
+  }
+
+  override def fix(implicit doc: SemanticDocument): Patch = {
+    lazy implicit val unit = global.newCompilationUnit(doc.input.text, doc.input.syntax)
+
+    val patchForExplicitResultTypes = addExplicitResultType()
+    val patchForTypeApply = AddTypeApply()
+
+    patchForExplicitResultTypes + patchForTypeApply
+  }
+
+  private def AddTypeApply()(implicit doc: SemanticDocument,
+                             unit: global.CompilationUnit): Patch = {
+    doc.synthetics.collect {
+      case syn@TypeApplyTree(function: SemanticTree, typeArguments: List[SemanticType]) =>
+        val symbols = typeArguments.map(getAbsoluteType)
+        if (symbols.length < typeArguments.length) Patch.empty
+        else {
+          val gsymbols = symbols.sequence
+          (for {
+            originalTree <- SyntheticHelper.getOriginalTree(syn)
+            if (!originalTree.isInstanceOf[Term.ApplyInfix])
+            if (!(originalTree.isInstanceOf[Pat] && !originalTree.isInstanceOf[Term])) // Never add types on elements that are on the right side of "="
+            replace <- if (syn.toString.startsWith("*.apply")) Some(".apply")
+            else if (syn.toString.startsWith("*[")) Some("")
+            else None
+            typesFromSemantic = gsymbols
+            typesSeenFromGlobal = getTypeNameAsSeenByGlobal(originalTree).map(_.map(_.dealiasWiden))
+            types <- typesSeenFromGlobal.orElse(typesFromSemantic)
+          } yield Patch.addRight(originalTree, s"${replace}[${types.mkString(", ")}]")).getOrElse(Patch.empty)
+        }
+    }.toList.reverse.asPatch
+  }
+
+
+  private def getTypeNameAsSeenByGlobal(origin: Tree)
+                                       (implicit doc: SemanticDocument,
+                                        unit: global.CompilationUnit): Option[List[global.Type]] = {
+    for {
+      globalOrigin <- CompilerService.getTreeInGlobal(origin.tokens.last, global)
+      types <- globalOrigin match {
+        case t@global.Apply(fun, args) => {
+          fun match {
+            case global.TypeApply(fun, args) =>
+              Some(args.map(_.tpe.asInstanceOf[global.Type]))
+            case _ => None
+          }
+        }
+        case _ => None
```

Comment:
I suggest to avoid nesting pattern matching when possible. It improves readability:

```suggestion
        case global.Apply(global.TypeApply(fun, args), _) =>
          Some(args.map(_.tpe.asInstanceOf[global.Type]))
        case _ => None
```

Reply:
nice. thanks